design/one-pager-host-aware-stack-manager.md

```diff
@@ -0,0 +1,211 @@
+ However, for this proposal, we need to authenticate two different Kubernetes API's inside the stack manager pod and
+ we can only use `in cluster config` for one of them and will take a `kubeconfig` as parameter for the other.
+ 
+Option 1 - Using `in cluster config` for Host Cluster: Host Cluster already auto-mounts service account tokens which we
+ can use.
+ 
+Option 2 - Using `in cluster config` for Tenant Kubernetes: This will be consistent with other controllers (Crossplane 
+ and stack controllers).
+ 
+To leverage existing mechanisms to authenticate to host cluster, we will use Option 1. We will define RBAC rules for
+ stack manager which will be part of `crossplane-controllers` helm chart (see [packaging changes](#packaging-changes)) 
+ that will be deployed into host cluster. This way, we can explicitly declare permissions of stack manager on Host
+ Cluster.
+
+#### Stack Installation
+
+In case of a stack installation, Stack Manager should:
+
+1. Create install job on Host Cluster and read logs of install pod from Host.
+2. Deploy CRDs and RBAC on Tenant Kubernetes
+3. Wait until a service account token secret is created for the stack to be installed.
+4. Copy service account token secret into Host Cluster.
+5. Deploy stack controller deployment by [configuring against Tenant Kubernetes Apiserver](#configuring-a-pod-against-tenant-kubernetes-api-server).
+
+
+##### Mapping Resources to Host
+
+While creating resources (token secrets or controller deployments) on Host Cluster, Stack Manager will map objects
+ (jobs, deployments, secrets ...) on multiple namespaces of Tenant Kubernetes into a single namespace on Host Cluster
+ by prepending `<namespace>.` to object name. For example:
+
+ | Tenant Kubernetes | Host Cluster |
+ | :---: | :---: |
+ | Namespace / Name | Namespace / Name |
+ | crossplane-system / stack-gcp | tenant-n-controllers / crossplane-system.stack-gcp |
+ | dev / stack-wordpress | tenant-n-controllers / dev.stack-wordpress |
+
+This mapping will cause a length limit to resource names as follows:
+
+```
+len(namespace) + len(resource-name) + 1 <= 253
+```
+
+Right now, this will effect:
+- Install job name
+- Stack controller deployment name
+- Stack controller service account name
+
+However, considering the following, this should not cause a new limitation for Crossplane:
+
+- There are already labels holding some resource names (e.g. `core.crossplane.io/parent-name: stack-gcp`) and Kubernetes
+ Labels are [limited to 63 chars](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)
+- Kubernetes namespaces are limited to 63 chars
+ 
+#### Stack Uninstallation
+
+Stack manager can no longer rely on owner references for deletion of stack install jobs and controller deployments
+ since stack related custom resources and jobs/deployments live in different Kubernetes Apiservers. Label based cleanup
+ will be implemented as a solution.
```

Comment:
I'm glad you pointed this out.

It will take some retooling or atleast renaming to handle the case where the parent resource resides in another (a host) cluster.

```yaml
kind: Deployment
metadata:
  name: tenant-controller
  namespace: tennant-namespace
  labels:
    core.crossplane.io/parent-group: stacks.crossplane.io
    core.crossplane.io/parent-version: v1alpha1
    core.crossplane.io/parent-kind: Stack
    core.crossplane.io/parent-name: tenant-stack
    core.crossplane.io/parent-namespace: tennant-n-stacks ?
    core.crossplane.io/parent-uid: dead-beef
    core.crossplane.io/parent-cluster: ?
```

When the Stack is deleted from the Host cluster, the `DeleteAllOf` will need to know the namespace of the tenant resources, so there will need to be some hinting that _hosted_ stack is special and the tenant namespace should be parsed from the stack name.

Reply:
Since we moved away from owner based deletion (which didn't work anyway), a stack-manager must act on the resources of the stack. I assume the Stack resource lives in the host cluster for this reason.

Reply:
I think you're assuming `Stack` lives in host cluster here. However, this is not the case as stated in previous comment (`Stack` lives in `Tenant Kubernetes Instance`) which simplifies and invalidates above mentioned problems.

Reply:
Indeed. https://github.com/crossplaneio/crossplane/pull/1138#discussion_r364281011